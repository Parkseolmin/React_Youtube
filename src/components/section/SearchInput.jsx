import { useCallback, useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';

export default function SearchInput() {
  // const [search, setSearch] = useState('');
  const searchRef = useRef('');
  const navigate = useNavigate();
  const handleChange = useCallback((e) => {
    searchRef.current = e.target.value;
    // setSearch(e.target.value);
  }, []);
  const handleSubmit = (e) => {
    e.preventDefault();
    const search = searchRef.current.trim();
    if (search.trim() !== '') {
      navigate(`/search/${search}`);
      // setSearch('');
      searchRef.current = ''; // clear input field after submit
      e.target.reset();
    }
  };

  return (
    <div id='search'>
      <div className='search__inner'>
        <form onSubmit={handleSubmit}>
          <label htmlFor='searchInput'>
            <span className='ir'>검색</span>
          </label>
          <input
            type='text'
            id='searchInput'
            // value={search}
            onChange={handleChange}
            className='search__input '
            placeholder='Search...'
          />
        </form>
      </div>
    </div>
  );
}
