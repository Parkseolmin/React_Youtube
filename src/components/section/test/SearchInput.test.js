import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Routes, Route, useNavigate } from 'react-router-dom';
import SearchInput from '../SearchInput';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('SearchInput Test', () => {
  it('renders SearchInput', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <SearchInput />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('updates input value on change', () => {
    render(
      <MemoryRouter>
        <SearchInput />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'test' } });

    expect(input.value).toBe('test');
  });

  it('navigates to search page on submit with valid input', () => {
    const mockNavigate = jest.fn();
    useNavigate.mockReturnValue(mockNavigate);

    render(
      <MemoryRouter>
        <SearchInput />
      </MemoryRouter>
    );

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'test' } });

    fireEvent.submit(screen.getByLabelText(/검색/i));

    // 네비게이션이 올바른 URL로 이루어졌는지 확인
    expect(mockNavigate).toHaveBeenCalledWith('/search/test');
  });

  it('입력 필드가 비어있을 때 navigate가 호출되지 않는지 테스트', () => {
    const mockNavigate = jest.fn();
    useNavigate.mockReturnValue(mockNavigate);

    render(
      <MemoryRouter>
        <SearchInput />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');

    // 빈 값으로 제출
    fireEvent.change(input, { target: { value: '' } });
    fireEvent.submit(screen.getByLabelText(/검색/i));

    expect(mockNavigate).not.toHaveBeenCalled();
  });
});
