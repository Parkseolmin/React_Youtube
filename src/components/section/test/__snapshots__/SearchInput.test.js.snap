// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Test renders SearchInput 1`] = `
<DocumentFragment>
  <div
    id="search"
  >
    <div
      class="search__inner"
    >
      <form>
        <label
          for="searchInput"
        >
          <span
            class="ir"
          >
            검색
          </span>
        </label>
        <input
          class="search__input "
          id="searchInput"
          placeholder="Search..."
          type="text"
          value=""
        />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
